### PR TITLE
JWT - Adds a custom dynamic secret fetcher that can read a tokens header

### PR DESCRIPTION
**Pull request by hassox**
Addresses: https://github.com/ueberauth/guardian/issues/469

### CODE REVIEW

**Comment by hassox** [comment]
@yordis updated

**Comment by hassox** [comment]
@yordis ping

**Comment by apognu** [comment]
Does `fetch_signing_secret` only returns the secret? Signing tokens with different secrets, but not being able to identify which one from the token defeats the purpose, somewhat.

`fetch_signing_secret` should return the secret **and** a key id that would appear in the generated token header as a ```kid``` claim.

Right now, there is no way, from the token headers passed to ```fetch_verifying_secret```, to fetch the key used to sign a token.

**Comment by scrogson** [comment]
@apognu in general, I agree with you. If you are doing public/private key signing/verifying and rotating keys then what you are saying makes perfect sense.

The default JWT token module that ships with Guardian is more naive and doesn't really go to great lengths to support multiple secret keys.

The great news is that you can easily create your own token strategy and use that instead of the default one provided by Guardian.

Let me know if we're missing something.

**Comment by apognu** [comment]
Thanks for replying on a closed issue.

I agree that the default behavior should be a naive one, and while Guardian does allow me to specify custom `secret_fetcher` and `token_verify_module` (if there is another plugging point, maybe I missed it), I believe they lack specific features that would be useful:

 * `fetch_verifying_secret` passes in the JWT headers, so we are able to look for the proper key
 * `fetch_signing_secret` has nothing to do with the JWT headers, so we cannot set, for instance, the `kid`

My specific use-case aside, I believe there is a symmetry issue between those two functions, where one can use the values in the headers and the second cannot set those values.

On the same note (but not directly related to this issue), in custom `token_verify_module`s, the token header is not passed to the `verify_claim` methods, so we still cannot map any verification rules to the key that is used (for example, "a token signed by a specific key can only include the `sub` to which the key belongs to").

**Comment by hassox** [comment]
@apognu I don't believe this has been released yet. I'd be interested to see what modifications you would put in place to make this feature more useful.

**Comment by scrogson** [comment]
@apognu I see. Excellent points! I think it would be great to support these things. Perhaps we can extend what's in master before the next release.

**Comment by apognu** [comment]
@hassox @scrogson I could give it a go, but by the time I familiarize myself with the structure of the project, it could be some time. When is the next release scheduled?

Without having looked too closely at the code, this would require:

 * passing the token headers left and right, so some or all headers are available at different steps of the processing (including in `token_verify_module`;
 * allowing the `secret_fetcher` to modify the header claims (or a subset thereof; should we allow this to dynamically determine the `alg`, on top of the `kid` and secret key?).

Those are my thoughts right now, but I have not spent a lot of time working on it. And I'll be frank, I have no idea how much work that would represent on Guardian's code base.

File: lib/guardian/token/jwt.ex [hunk 1]
```diff
@@ -124,6 +146,66 @@ defmodule Guardian.Token.Jwt do
...
+          do: DI.fetch_verifying_secret(mod, token_headers, opts)
+
+        defoverridable fetch_signing_secret: 2, fetch_verifying_secret: 3
+      end
+    end
+  end
+
+  defmodule SecretFetcher.SecretFetcherDefaultImpl do
+    @moduledoc false
+    use Guardian.Token.Jwt.SecretFetcher
```
Comment (ps-feng) [inline lib/guardian/token/jwt.ex]: I think it's possible to get rid of the macro altogether, but please excuse me if this makes no sense as I've just started learning Elixir and using Guardian to verify Firebase tokens...

Isn't it enough with `@behaviour Guardian.Token.Jwt.SecretFetcher`? After all `fetch_secret_fetcher` is already referencing `SecretFetcherDefaultImpl` by default. This way we can get rid of the macro, which also resembles OO inheritance.

**Pull request merged**